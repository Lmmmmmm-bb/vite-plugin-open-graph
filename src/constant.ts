export const EXCEPTION_FIELD = ['secureUrl', 'siteName']
