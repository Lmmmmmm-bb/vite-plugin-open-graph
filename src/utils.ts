import type { HtmlTagDescriptor } from 'vite'
import { EXCEPTION_FIELD } from './constant'

/**
 * transform `siteName` to `site:name`
 * @param attrName name
 * @returns join by ':'
 */
export const camelcase = (attrName: string) => attrName.replace(/([A-Z])/g, EXCEPTION_FIELD.includes(attrName) ? '_$1' : ':$1').toLowerCase()

export const toAttrs = (name: string, content: number | string | string | Record<string, string | number>, field: 'name' | 'property', prefix?: string): HtmlTagDescriptor['attrs'] | HtmlTagDescriptor['attrs'][] => {
  const _name = prefix ? `${prefix}:${name}` : name

  if (typeof content === 'number')
    return { [field]: _name, content: `${content}` }

  else if (Array.isArray(content))
    return content.map(_content => ({ [field]: _name, content: _content }))

  else if (typeof content === 'object')
    return Object.entries(content).map(([key, value]) => toAttrs(camelcase(key), value, field, _name)) as HtmlTagDescriptor['attrs'][]

  return { [field]: _name, content }
}
