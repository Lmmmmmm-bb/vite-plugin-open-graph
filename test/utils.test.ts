import { describe, expect, it } from 'vitest';

import { ogOptions } from './constant.ts';
import { camelcase, toAttrs } from '../src/utils.ts';

describe('camelcase function', () => {
  it('lowercase name', () => {
    expect(camelcase('hi')).toEqual('hi');
    expect(camelcase('name')).toEqual('name');
  });

  it('normal name', () => {
    expect(camelcase('helloWorld')).toEqual('hello:world');
    expect(camelcase('localeAlternate')).toEqual('locale:alternate');
    expect(camelcase('playerWidth')).toEqual('player:width');
  });

  it('exception field name', () => {
    expect(camelcase('secureUrl')).toEqual('secure_url');
    expect(camelcase('siteName')).toEqual('site_name');
    expect(camelcase('appId')).toEqual('app_id');
  });

  it('empty string', () => {
    expect(camelcase('')).toEqual('');
  });
});

describe('toAttrs function', () => {
  it('snapshot', () => {
    const basicOGMetaAttrs = Object.entries(ogOptions.basic || {}).map(([name, content]) => toAttrs(camelcase(name), content, 'property', 'og'));
    const twitterOGMetaAttrs = Object.entries(ogOptions.twitter || {}).map(([name, content]) => toAttrs(camelcase(name), content, 'name', 'twitter'));

    expect(basicOGMetaAttrs).toMatchSnapshot();
    expect(twitterOGMetaAttrs).toMatchSnapshot();
  });
});
