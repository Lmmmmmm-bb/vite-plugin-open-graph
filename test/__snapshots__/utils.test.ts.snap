// Vitest Snapshot v1

exports[`toAttrs function > snapshot 1`] = `
[
  {
    "content": "https://lmmmmmm.me",
    "property": "og:url",
  },
  {
    "content": "_lmmmmmm",
    "property": "og:title",
  },
  {
    "content": "image.png",
    "property": "og:type",
  },
  {
    "content": "https://lmmmmmm.me/avatar.png",
    "property": "og:image",
  },
  {
    "content": "auto",
    "property": "og:determiner",
  },
  {
    "content": "_lmmmmmm, Front-end Developer.",
    "property": "og:description",
  },
  {
    "content": "zh_CN",
    "property": "og:locale",
  },
  [
    {
      "content": "fr_FR",
      "property": "og:locale:alternate",
    },
    {
      "content": "es_ES",
      "property": "og:locale:alternate",
    },
  ],
  {
    "content": "_lmmmmmm",
    "property": "og:site_name",
  },
  [
    {
      "content": "audio url",
      "property": "og:audio:url",
    },
    {
      "content": "audio secure url",
      "property": "og:audio:secure_url",
    },
    {
      "content": "video.movie",
      "property": "og:audio:type",
    },
  ],
  {
    "content": "video meta",
    "property": "og:video",
  },
]
`;

exports[`toAttrs function > snapshot 2`] = `
[
  {
    "content": "https://lmmmmmm.me/avatar.png",
    "name": "twitter:image",
  },
  {
    "content": "twitter image alt",
    "name": "twitter:image:alt",
  },
  {
    "content": "player",
    "name": "twitter:player",
  },
  {
    "content": "1200",
    "name": "twitter:player:width",
  },
  {
    "content": "600",
    "name": "twitter:player:height",
  },
  {
    "content": "player stream",
    "name": "twitter:player:stream",
  },
  [
    [
      {
        "content": "iphone name",
        "name": "twitter:app:name:iphone",
      },
      {
        "content": "ipad name",
        "name": "twitter:app:name:ipad",
      },
      {
        "content": "google play name",
        "name": "twitter:app:name:googleplay",
      },
    ],
    [
      {
        "content": "iphone url",
        "name": "twitter:app:id:iphone",
      },
      {
        "content": "ipad url",
        "name": "twitter:app:id:ipad",
      },
      {
        "content": "google play url",
        "name": "twitter:app:id:googleplay",
      },
    ],
    [
      {
        "content": "iphone url",
        "name": "twitter:app:url:iphone",
      },
      {
        "content": "ipad url",
        "name": "twitter:app:url:ipad",
      },
      {
        "content": "google play url",
        "name": "twitter:app:url:googleplay",
      },
    ],
  ],
]
`;
